### ISSUE TITLE
macOS: Ignore .DS_Store files

### ISSUE THREAD

**Issue by autoantwort** [microsoft/vcpkg#16520]
**Describe the bug**
For example when I install the package `icu`, I get the following error:
```
...
-- Performing post-build validation
The following files are placed in
/Users/leanderSchulten/git_projekte/Lichtsteuerung/vcpkg/packages/icu_arm64-osx:

    /Users/leanderSchulten/git_projekte/Lichtsteuerung/vcpkg/packages/icu_arm64-osx/.DS_Store

Files cannot be present in those directories.
```

**Environment**
- OS: macOS

**To Reproduce**
1. ./vcpkg install icu

**Expected behavior**
The package should install without problems 

**Additional context**
That happens not only with this package

**Comment by autoantwort** [comment]
What I tried: Delete the package/icu and the buildtree/icu folder and build with `--binarysource=clear`, the error does not go away

**Comment by JackBoosY** [comment]
It seems to be a recovery file created by osx, @strega-nil can you please take a look？

Thanks.

**Comment by autoantwort** [comment]
That are files created by the Finder so save the "view" settings (e.g. in this folder sort the files by size). [Wikipedia](https://en.wikipedia.org/wiki/.DS_Store)

**Comment by BillyONeal** [comment]
It seems like the right behavior is actually to delete these rather than ignore them?

**Comment by autoantwort** [comment]
> It seems like the right behavior is actually to delete these rather than ignore them?

Yeah, but the finder would recreate them after you delete them.

**Comment by BillyONeal** [comment]
Why is finder creating them in the first place? (We aren't seeing this behavior on our local macs or in our build fleet)

**Comment by autoantwort** [comment]
Because I opened the `buildtree` folder in the finder to see the error logs

**Comment by BillyONeal** [comment]
> Because I opened the `buildtree` folder in the finder to see the error logs

You would need to be opening the `packages` folder to see the behavior above, and you would need to be doing so during the build (because we already delete recursive on any packages folder that is about to be used for install).

**Comment by autoantwort** [comment]
Yeah I open the file tree, then you also see the packages folder and it maybe also create files there. But I definitely get this error from time to time when I open the vcpkg folder in finder. 
Or I get errors like this:
```
The following files are already installed in /Users/leanderSchulten/git_projekte/Lichtsteuerung/vcpkg_installed/x64-osx and are in conflict with pcre2:x64-osx

Installed by libpq:x64-osx
    .DS_Store
```

**Comment by BillyONeal** [comment]
To be clear: I'm not saying that we shouldn't skip these. But I am saying that anything that is messing with a packages tree in the window between when it is produced by a build and when we install it is a serious problem, and we need to have a firm understanding of why that is happening. For example, correct function of binary caching assumes that a given packages tree is only a function of inputs that vcpkg and the build systems it invokes put there, not of anything messing with the packages tree after the fact.

To that end, I think we really need to understand why/how this is happening before taking action.

* If you only get to this place by someone digging around in the packages tree during construction of the packages tree, then the user is doing something we can't really support, which argues to add a post build lint check to always fail the build if .DS_Store is ever seen.
* On the other hand, if 'normal user behavior' can trigger the issue, that argues to add a mechanism to skip those.

Hence really needing a repro

**Comment by autoantwort** [comment]
> Hence really needing a repro

I don't know what you need, but I have created a small demo. I build a port with errors. Opens the package folder in the finder and move some folders up, "fix" the error (and have trouble with binary caching ;) ), build again and then get the `.DS_Store` error. 

https://user-images.githubusercontent.com/41973254/127757149-7fe82d82-e810-4e15-b7c6-369c3685a044.mp4

**Comment by autoantwort** [comment]
This is especially annoying when you build for example qtbase for 25 minutes and then there is a .DS_Store file and you can wait another 25 minutes ...

**Comment by BillyONeal** [comment]
> > Hence really needing a repro
> 
> I don't know what you need, but I have created a small demo. I build a port with errors. Opens the package folder in the finder and move some folders up, "fix" the error (and have trouble with binary caching ;) ), build again and then get the `.DS_Store` error.

Right, the question is "why does this happen to you but we have not been able to make it happen ourselves." I agree with you that it's an important problem, I'm only worried about hiding other types of behavior we can't support (like mucking around in the packages tree manually before/during install).

> It happens because I open the packages folder (for example after a build failure), a hidden `.DS_Store` file is created, then I leave the folder. Now I fix the port error and build again, vcpkg deletes the package folder and recreates it, the finder sees that the `.DS_Store` file is now missing and creates it. The port builds now successfully, but now I get the error that there is a `.DS_Store` file in the package folder.

I see, if finder is open to a directory and it's deleted, finder remains open to the now gone directory? Maybe the right fix is to put a sha in the packages folder name like we do for the sources...

> PS: I got this error today again, was a good reminder to answer here :D

Thanks for following up!

**Comment by autoantwort** [comment]
> why does this happen to you but we have not been able to make it happen ourselves.

Have you tried it locally? 

> I agree with you that it's an important problem, I'm only worried about hiding other types of behavior we can't support

I thought of simply removing the .DS_Store files (finder needs some seconds (>30?) to recreate the files. When I remove them with `rm` they never come back) or ignoring them. I don't think that this can break anything or enable stuff that should not be possible. 

> I see, if finder is open to a directory and it's deleted, finder remains open to the now gone directory?

Like in the video I can even leave the folder and so you don't see that folder anymore and you get this behavior.

> Maybe the right fix is to put a sha in the packages folder name like we do for the sources...

That must then be a random string. But then this pollute everything and you have hundreds of folders that should be deleted.

**Comment by BillyONeal** [comment]
> Have you tried it locally?

Yep, never had this happen. But also never opened finder into packages land.

> > Maybe the right fix is to put a sha in the packages folder name like we do for the sources...
> 
> That must then be a random string. But then this pollute everything and you have hundreds of folders that should be deleted.

I mean, that's the status quo. It isn't like the packages tree is usable after things are mapped into installed.

**Comment by autoantwort** [comment]
> But also never opened finder into packages land.

Yeah ok then this can not happen :D

Can we not simply skip `.DS_Store` files in `get_files_impl`?

**Comment by BillyONeal** [comment]
> Can we not simply skip `.DS_Store` files in `get_files_impl`?

It needs to be a bit more than that because sometimes we pass whole directories to tools (e.g. 7z / zip / unzip / tar) but that's mostly it.

**Comment by autoantwort** [comment]
Now it fails to install with a conflict:
```
Installing package qtdeclarative[core]:x64-osx...
The following files are already installed in /Users/leanderSchulten/git_projekte/Lichtsteuerung/vcpkg_installed/x64-osx and are in conflict with qtdeclarative:x64-osx

Installed by qtbase:x64-osx
    .DS_Store
    include/.DS_Store
```

**Comment by autoantwort** [comment]
I now also often get the following error:
```
➜  vcpkg git:(tcl-no-absolute) ✗ ./vcpkg x-add-version --all && git add versions && git commit --amend --no-edit
Found both manifest and CONTROL file in port /Users/leanderSchulten/git_projekte/Lichtsteuerung/vcpkg/ports/.DS_Store; please rename one or the other
```
